### PR TITLE
feat(tooltip): html updates for a11y

### LINKED ISSUE #3100: Tooltip: AVT 1 violations
## Environment

> Operating system

Mac

> Browser

Chrome, Safari

> Automated testing tool and ruleset

AAT 4.2.1

> Assistive technology used to verify

VoiceOver

## Detailed description

Tooltip has multiple AAT violations

> What version of the Carbon Design System are you using?

10.3.1

> What did you expect to happen?

There to not be AAT violations

> What happened instead?

3 violations (color contrast and landmarks)

> What WCAG 2.1 checkpoint does the issue violate?

See screen shot

## Steps to reproduce the issue

1. Force the tooltip to be open by adding the `bx--tooltip--shown` class (trying to run DAP with default behaviors I couldn't get the tooltip to remain open while running the tests)
2. Run DAP

## Additional information

I would expect the dialog to have `role="dialog"` not the trigger (solve the violation about the landmark) while the trigger uses `aria-controls` attribute to create the relationship between the two elements. 

`role="tooltip"` should be be for static text only (i.e. definition tooltip #3099)

Screen Shot 2019-06-18 at 1 20 29 PM

### PR DESCRIPTION
**Pull request by elizabethsjudd**
Closes #3100 

Simple HTML updates to pass AVT 1.  The tooltip still has some bugs for keyboard and screen reader users (AVT 2 and AVT 3) that I can take a look at next but it's going to require some updates to the JavaScript as well as getting an answer to my [open question](https://github.com/carbon-design-system/carbon/issues/3100#issuecomment-504550259)

#### Changelog

**New**

- Add `role="dialog"` to define a tooltip with interactive elements (remove AVT 1 landmark violation)
- Add `aria-controls`, `aria-haspopup` to create the relationship between the trigger and tooltip elements
- Add demos for various ways users can associate the label for the tooltip element

**Changed**

- Update trigger `role="tooltip"` to `role="button"`

#### Testing / Reviewing

- Regression test that styling/functionality remained the same

### CODE REVIEW

File: packages/components/src/components/tooltip/tooltip.hbs [hunk 1]
```diff
@@ -5,17 +5,21 @@
...
     {{ carbon-icon 'Information16' }}
   </div>
 </div>
+{{!-- @todo add aria-hidden once JavaScript can toggle it --}}
 <div id="unique-tooltip" data-floating-menu-direction="bottom" class="{{@root.prefix}}--tooltip"
-  data-avoid-focus-on-open>
+  data-avoid-focus-on-open role="dialog" aria-describedby="unique-tooltip-body" aria-labelledby="tooltip-label">
   <span class="{{@root.prefix}}--tooltip__caret"></span>
-  <p>This is some tooltip text. This box shows the maximum amount of text that should appear inside. If more room is
+  <p id="unique-tooltip-body" >This is some tooltip text. This box shows the maximum amount of text that should appear inside. If more room is
```
Comment (asudoh) [inline packages/components/src/components/tooltip/tooltip.hbs]: Could you generate unique ID in `.config.js` and refer to that here? Thanks!

Comment (scottnath) [reply packages/components/src/components/tooltip/tooltip.hbs]: Hi @asudoh @elizabethsjudd  could these be UUID? [node-uuid](https://github.com/kelektiv/node-uuid) could be used for this, generating unique keys for each demo (so they don't override each other.) Would require a new dependency, but node-uuid is the most popular/supported tool for this.

**Review by jnm2377 — APPROVED** [review]
Working as expected for me. Thanks for this! 👍

File: packages/components/src/components/tooltip/tooltip.hbs (hunk 1 above)
Comment (asudoh) [reply packages/components/src/components/tooltip/tooltip.hbs]: @scottnath Any solutions generating unique ID is fine. e.g. This is our dev-time only I wouldn't foresee a collision with `Math.random()`, etc.

Comment (scottnath) [reply packages/components/src/components/tooltip/tooltip.hbs]: @asudoh confused by the message. Does that mean it's ok to add the package [node-uuid](https://github.com/kelektiv/node-uuid) and to use it to create ID's with uuid? This package would then be available to any component that needs to create unique ID's

**Comment by joshblack** [comment]
Could you share the AVT1 violations and how to re-create them for the interactive tooltips?

Definitely agreed that interactive tooltips should be replaced by dialog since they have interactive content, however it seems like this PR relies on some updates to the JS to make things fully accessible/functional?

**Review by joshblack — COMMENTED** [review]
Could you share the AVT1 violations and how to re-create them for the interactive tooltips?

Definitely agreed that interactive tooltips should be replaced by dialog since they have interactive content, however it seems like this PR relies on some updates yet to be made to the JS to make things fully accessible/functional?

**Comment by elizabethsjudd** [comment]
@joshblack The AVT 1 violation is that the tooltip didn't have a role so the content wasn't in a landmark.

In order to recreate the violation:
1. Open a tooltip
2. force the tooltip open by adding `bx--tooltip--shown`
3. run DAP. (DAP doesn't run on hidden content and turning on that setting I've been told is buggy)

You'll still get a color contrast violation with the link but that is design specific so I wasn't focused on it for this PR until @carbon-design-system/design has a chance to look at it. 

You are correct that it will require some JS updates to make it fully accessible which is what I plan on working next.  I do need this [question](https://github.com/carbon-design-system/carbon/issues/3100#issuecomment-504550259) answered by @carbon-design-system/design before I work on it though.  This PR resolves the HTML AVT 1 violation but also doesn't make the component less a11y than the current version as the current one is also inaccessible for screen reader users.

**Comment by joshblack** [comment]
@elizabethsjudd this is what I see when running it with the tooltip shown:

image

And unfortunately none of the violations relate to the tooltip node, is there a step I'm missing here?

To help with debugging, here are the rulesets I have enabled:

image

**Comment by elizabethsjudd** [comment]
@joshblack those are the correct violations you should be receiving when you run it on my updates. If you run DAP on the current version (not my branch) you will bet an additional violation for the tooltip node.

**Comment by joshblack** [comment]
Oh duh 😂 You're totally right, thanks @elizabethsjudd! Sorry for being dumb here haha

**Review by dakahn — APPROVED** [review]
Excellent! It's all clear here on my end. Thanks for contributing. 🌟

File: packages/components/src/components/tooltip/tooltip.hbs (hunk 1 above)
Comment (asudoh) [reply packages/components/src/components/tooltip/tooltip.hbs]: Using `node-uuid` is OK, but I wouldn't given there is an easier way. Also if you are strong for `node-uuid` please use it with `devDependencies`.

**Comment by jnm2377** [comment]
cc: @snidersd any comments before merging?

**Comment by snidersd** [comment]
@jnm2377 I'm not seeing any change when testing React https://deploy-preview-3148--carbon-components-react.netlify.com. The contrast in the pop-up still fails and there is one DAP issue. 
Screen Shot 2019-07-01 at 9 02 38 AM
 Should I be looking at a different link?

**Comment by elizabethsjudd** [comment]
@snidersd This issue ONLY applies to Carbon Vanilla.  WH doesn't use the React versions of Carbon.  There is still the color contrast on the link element but we need design input for that violation.  My goal was to resolve the other violations/HTML structure issues with the tooltip.  There will be additional JS updates needed to fully be accessible.

**Comment by snidersd** [comment]
@elizabethsjudd I understand that. I was sent a request to review this issue and just wanted to make sure that both Vanilla and React stay in sync since @dakahn is working on the React version.

**Comment by asudoh** [comment]
🎉 Thanks @elizabethsjudd for your contribution!

**Pull request merged**